cmake/functions/four_c_auto_define_tests.cmake

```diff
@@ -0,0 +1,230 @@
+# This file is part of 4C multiphysics licensed under the
+# GNU Lesser General Public License v3.0 or later.
+#
+# See the LICENSE.md file in the top-level for license information.
+#
+# SPDX-License-Identifier: LGPL-3.0-or-later
+
+function(_set_up_unit_test_target _target)
+  set(options "")
+  set(oneValueArgs NP THREADS)
+  set(multiValueArgs "")
+  cmake_parse_arguments(
+    _parsed
+    "${options}"
+    "${oneValueArgs}"
+    "${multiValueArgs}"
+    ${ARGN}
+    )
+
+  if(DEFINED _parsed_UNPARSED_ARGUMENTS)
+    message(SEND_ERROR "There are unparsed arguments: ${_parsed_UNPARSED_ARGUMENTS}")
+  endif()
+
+  if(NOT DEFINED _parsed_NP)
+    set(_parsed_NP 1)
+  endif()
+
+  if(NOT DEFINED _parsed_THREADS)
+    set(_parsed_THREADS 1)
+  endif()
+
+  set(assert_mpi_file ${CMAKE_CURRENT_BINARY_DIR}/assert_mpi_${_target}_${_parsed_NP}.cpp)
+  set(FOUR_C_ADD_GOOGLE_TEST_EXECUTABLE_NP ${_parsed_NP})
+  configure_file(${PROJECT_SOURCE_DIR}/unittests/assert_mpi.cpp.in ${assert_mpi_file})
+
+  message(VERBOSE "Setting up unit test target ${_target}")
+
+  add_executable(
+    ${_target}
+    ${PROJECT_SOURCE_DIR}/unittests/4C_gtest_main_mpi_test.cpp ${assert_mpi_file} ${_parsed_SOURCE}
+    )
+  # Store unit test executables directly inside the tests/ directory
+  set_target_properties(${_target} PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/tests)
+
+  # Do not try to build tests as unity files.
+  set_target_properties(${_target} PROPERTIES UNITY_BUILD OFF)
+
+  # All libraries are linked as PRIVATE since a unit test executable cannot be used as a dependency itself.
+
+  # Common dependencies for unit tests
+  target_link_libraries(${_target} PRIVATE four_c_private_compile_interface)
+  target_link_libraries(${_target} PRIVATE gtest gmock)
+  target_link_libraries(${_target} PRIVATE unittests_common)
+
+  # Link the main library
+  # NOTE: We can think about linking a subset of classes under test here. The module object targets would be a good
+  # candidate for this. However, as long as the interdependencies between modules are high, this is not feasible.
+  target_link_libraries(${_target} PRIVATE ${FOUR_C_LIBRARY_NAME})
+
+  # the first process will write a unit test report
+  separate_arguments(
+    MPIEXEC_EXTRA_OPTS_FOR_TESTING_LIST UNIX_COMMAND ${MPIEXEC_EXTRA_OPTS_FOR_TESTING}
+    )
+
+  set(mpi_arguments
+      ${MPIEXEC_EXTRA_OPTS_FOR_TESTING_LIST}
+      -np
+      1
+      $<TARGET_FILE:${_target}>
+      --gtest_output=xml:unittest_reports/${_target}_report.xml
+      )
```

Comment:
We could really think about our cmake formatting settings and increase the line length 😅